_posts/2022-04-24-pyflink-1.15-thread-mode.md

```diff
@@ -0,0 +1,463 @@
+      <td style="text-align: center">200ns</td>
+      <td style="text-align: center">10ms</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Process Mode (100k)</td>
+      <td style="text-align: center">900</td>
+      <td style="text-align: center">5s-10s</td>
+      <td style="text-align: center">5s</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Process Mode (10k)</td>
+      <td style="text-align: center">7000</td>
+      <td style="text-align: center">5s-10s</td>
+      <td style="text-align: center">3s</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Process Mode (1k)</td>
+      <td style="text-align: center">3.6w</td>
+      <td style="text-align: center">3s</td>
+      <td style="text-align: center">3s</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Process Mode (100)</td>
+      <td style="text-align: center">12w</td>
+      <td style="text-align: center">2s</td>
+      <td style="text-align: center">2s</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Thread Mode (100k)</td>
+      <td style="text-align: center">1200</td>
+      <td style="text-align: center">1ms</td>
+      <td style="text-align: center">100ms</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Thread Mode (10k)</td>
+      <td style="text-align: center">1.2w</td>
+      <td style="text-align: center">20us</td>
+      <td style="text-align: center">10ms</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Thread Mode (1k)</td>
+      <td style="text-align: center">5w</td>
+      <td style="text-align: center">3us</td>
+      <td style="text-align: center">10ms</td>
+    </tr>
+    <tr>
+      <td style="text-align: center">Thread Mode (100)</td>
+      <td style="text-align: center">12w</td>
+      <td style="text-align: center">1us</td>
+      <td style="text-align: center">10ms</td>
+    </tr>
+  </tbody>
+</table>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2022-04-24-pyflink-1.15-thread-mode/pyflink-performance.jpg"/>
+</center>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2022-04-24-pyflink-1.15-thread-mode/pyflink-latency.jpg"/>
+</center>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2022-04-24-pyflink-1.15-thread-mode/pyflink-checkpoint-time.jpg"/>
+</center>
+
+As we can see from the test results:
+
+ * If you care about latency and checkpoint time, thread mode is your better choice than process mode. The processing latency is decreased from several seconds in process mode to microseconds in thread mode. 
+
+ * Thread mode can bring better performance than process mode when data serialization/deserialization is not negligible relative to UDF calculation itself.
+Compared to process mode, benchmark has shown that the throughput could be increased by 2x in common scenarios such as json processing in thread mode.
+​​On the contrary, it is more recommended to use process mode, because different processes have better resource isolation.
```

Comment:
```suggestion
However, if the UDF calculation is slow and spends much longer time, then it is more recommended to use process mode, because the process mode is more mature and it has better resource isolation.
```